### PR TITLE
Payload store refactors

### PR DESCRIPTION
**Pull request by NiloCK**
PR:

1. Includes a few small refactors to the payloadStore.
    - removes its mutex (safety provided further up the chain)
    - removes its interface (interface abstraction is overkill for now)
    - removes some redundant logic
2. Adds some functionality to the payloadStore
    - `Remove`, `Clear`
3. Fixes existing functionality
    - rollback now rolls back rather than clears the store.
4. Adds tests for the above.

## Summary by CodeRabbit

- **New Features**
	- Enhanced payload management capabilities including the ability to remove, rollback, and clear payloads.
  
- **Bug Fixes**
	- Improved logic in payload addition and current payload retrieval functions for better accuracy and performance.

- **Refactor**
	- Updated payload store interface and method signatures for streamlined operations.

- **Tests**
	- Added and modified tests to ensure robust handling of payloads, including testing new functionalities like clearing and batch removal based on specific criteria.

### CODE REVIEW

File: app/peptide/payloadstore/payloadstore.go [hunk 1]
```diff
@@ -1,70 +1,100 @@
...
 	heights  map[int64]engine.PayloadID
 	current  *monomer.Payload
 }
 
-var _ PayloadStore = (*pstore)(nil)
-
-func NewPayloadStore() PayloadStore {
-	return &pstore{
-		mutex:    sync.Mutex{},
+func NewPayloadStore() *Store {
```
Comment (joshklop) [inline app/peptide/payloadstore/payloadstore.go:16]: nit: `payloadstore.NewPayloadStore` stutters. Let's use `New` instead.

File: app/peptide/payloadstore/payloadstore.go [hunk 2]
```diff
@@ -1,70 +1,100 @@
...
-type PayloadStore interface {
-	Add(payload *monomer.Payload)
-	Get(id engine.PayloadID) (*monomer.Payload, bool)
-	Current() *monomer.Payload
-	RollbackToHeight(height int64) error
-}
-
-type pstore struct {
-	mutex    sync.Mutex
+type Store struct {
```
Comment (joshklop) [inline app/peptide/payloadstore/payloadstore.go:10]: Let's add a comment noting that `Store` is not goroutine-safe.

File: app/peptide/payloadstore/payloadstore.go [hunk 3]
```diff
@@ -1,70 +1,100 @@
...
+func (p *Store) Add(payload *monomer.Payload) {
 	id := payload.ID()
-	p.mutex.Lock()
-	defer p.mutex.Unlock()
 	if _, ok := p.payloads[*id]; !ok {
 		p.heights[payload.Height] = *id
 		p.payloads[*id] = payload
-		p.current = payload
+
+		if (p.current == nil) || (payload.Height > p.current.Height) {
```
Comment (joshklop) [inline app/peptide/payloadstore/payloadstore.go:31]: The current payload isn't necessarily the payload with the greatest height. It is the payload that will become the next block.

File: app/peptide/payloadstore/payloadstore.go [hunk 4]
```diff
@@ -1,70 +1,100 @@
...
 	}
-	return nil, false
 }
 
-func (p *pstore) Current() *monomer.Payload {
-	p.mutex.Lock()
-	defer p.mutex.Unlock()
+func (p *Store) Get(id engine.PayloadID) (*monomer.Payload, bool) {
+	payload, ok := p.payloads[id]
+	return payload, ok
```
Comment (joshklop) [inline app/peptide/payloadstore/payloadstore.go:60]: Is it possible to just return `*monomer.Payload` here? If it's nil, we know it doesn't exist.

File: app/peptide/payloadstore/payloadstore.go [hunk 5]
```diff
@@ -1,70 +1,100 @@
...
+		if h > height {
+			delete(p.payloads, id)
+			delete(p.heights, h)
+		} else if h > max {
+			max = h
+		}
+	}
+
+	// set the current payload to the highest height
+	p.current = p.payloads[p.heights[max]]
```
Comment (joshklop) [inline app/peptide/payloadstore/payloadstore.go:85]: Similar to above. The current payload is the "next block", not the "highest block."

File: app/peptide/payloadstore/payloadstore.go [hunk 6]
```diff
@@ -1,70 +1,100 @@
...
-		return payload, true
+		delete(p.payloads, id)
+		delete(p.heights, payload.Height)
+	}
+
+	if resetCurrent {
+		p.current = nil
+		for _, payload := range p.payloads {
+			if p.current == nil || payload.Height > p.current.Height {
+				p.current = payload
```
Comment (joshklop) [inline app/peptide/payloadstore/payloadstore.go:52]: Same as above.

**Review by joshklop — COMMENTED** [review]
Made a first pass.

File: app/peptide/payloadstore/payloadstore.go (hunk 1 above)
Comment (NiloCK) [reply app/peptide/payloadstore/payloadstore.go:16]: 45518ed

File: app/peptide/payloadstore/payloadstore.go (hunk 2 above)
Comment (NiloCK) [reply app/peptide/payloadstore/payloadstore.go:10]: 45518ed

File: app/peptide/payloadstore/payloadstore.go (hunk 4 above)
Comment (NiloCK) [reply app/peptide/payloadstore/payloadstore.go:60]: 4371682

**Pull request closed without merging**